### ISSUE TITLE
Point-free `.parse` not possible since v3.9.0 – a feature or a bug?

### ISSUE THREAD

**Issue by christianhg** [colinhacks/zod#808]
Hi! 

Zod is an awesome project. And for a while I had code looking like this when parsing data from API endpoints:

```ts
import { z } from "zod";

type User = z.infer<typeof User>;
const User = z.object({
  name: z.string()
});

function getUser(): Promise<User> {
  return getUserFromApi().then(User.parse);
}

getUser()
  .then(
    // Use User
  )
  .catch(
    // Handle parsing error
  );
```

This is just a simplified example, but the idea is that `User.parse` is called by `.then` to parse whatever data is coming from the Promise. If `User.parse` throws an error, the whole thing results in a rejected Promise. Nice and simple and without the need to write `.then(user => User.parse(user))`.

However, since the release of v3.9.0, `.parse` cannot be used in a point-free way. An error is thrown inside Zod:

```
TypeError: Cannot read properties of undefined (reading 'safeParse')
```

Now, I want to ask if this behaviour is intentional or if it's a bug?

Disclaimers:
- I know that many people discourage point-free programming in JavaScript. It can be problematic, as the error above demonstrates, but I still happen to use the technique when I find that it simplifies my code.
- It's absolutely fine with me if this is not considered a bug. It's a small thing, and I'll keep using Zod even if I can't use `.parse` in a point-free way. I just wanted to make sure that I reported my findings.

I created a CodeSandbox that shows running code in v3.8.2, but if you change the version to v3.9.0 under "Dependencies", the example breaks: https://codesandbox.io/s/point-free-zod-parse-473q1?file=/src/index.ts

<img width="303" alt="Screen Shot 2021-12-02 at 22 23 08" src="https://user-images.githubusercontent.com/4602382/144505233-0c0597bb-adaf-4e3b-af32-7cbb07df2c26.png">

Finally, thank you for Zod. It's truly awesome 🙏

**Comment by JacobWeisenburger** [comment]
in v3.8.0, `parse` is an arrow function
[`parse: (data: unknown, params?: Partial<ParseParamsNoData>) => Output = (`](https://github.com/colinhacks/zod/blob/7fe864e34294b465362ba55aa59626208edb1a49/src/types.ts#L109)

but in v3.9.0 it was changed to a instance method
[`parse(data: unknown, params?: Partial<ParseParamsNoData>): Output {`](https://github.com/colinhacks/zod/blob/6aef50883d58acf5faea6ff7bdb88653109fb32d/src/types.ts#L168)

I'm pretty sure this change caused your issue. Now `this` is bound to a different context inside `parse`, so I'm pretty sure you will have to change your code to:

```ts
function getUser(): Promise<User> {
  return getUserFromApi().then( data => User.parse( data ) );
}
```
There might be a way to use the bind method to fix the issue, but personally I think using the bind method causes more confusion than just using an arrow function.

**Comment by christianhg** [comment]
Thanks for the clarification about what happened between v3.8 and v3.9. Nice job diving into this and figuring it out. It looks like a deliberate choice to move from arrow functions to methods, but the reason isn't clear to me: d47eadc014b1e488576e303f3ae9241f6b573a04

So my original question still stands: Is it a bug that `.parse` cannot be used in a point-free manner?

**Comment by JacobWeisenburger** [comment]
it looks like @colinhacks was the one to make that commit.

@colinhacks would you be able to weigh in on this?

**Comment by akomm** [comment]
I had to refactor code too after upgrading, due to the context used now.

`parse` has also the issue of accepting optional extra arguments, which makes it incompatible with assignment, where extra arguments are optional as well. The mismatch is due to caller potentially passing extra arguments of unexpected type. Had `parse` one argument, because it does not care about the caller extra arg, it would work.

**Comment by JacobWeisenburger** [comment]
Is this still an issue in the latest version v3.12.0?

**Comment by christianhg** [comment]
> Is this still an issue in the latest version v3.12.0?

I tried setting the version of `zod` to `3.12.0` in [the sandbox](https://codesandbox.io/s/point-free-zod-parse-473q1?file=/src/index.ts), and it seems to still be an issue.

**Comment by JacobWeisenburger** [comment]
Thanks I'll keep this open

**Comment by colinhacks** [comment]
Bug! Fixed in 3.12.1 - sorry about that.
